src/network/lan.cpp

```diff
@@ -0,0 +1,335 @@
+const static unsigned short int adv_port = 29998;
+static std::string ask_str;
+
+bool use_ipv6 = true;
+
+lan_adv::lan_adv() : Thread("lan_adv")
+{
+}
+
+void lan_adv::ask()
+{
+	if (!isRunning()) start();
+
+	if (ask_str.empty()) {
+		Json::Value j;
+		j["cmd"] = "ask";
+		ask_str = fastWriteJson(j);
+	}
+
+	send_string(ask_str);
+}
+
+void lan_adv::send_string(const std::string &str)
+{
+	if (g_settings->getBool("enable_ipv6")) {
+		std::vector<uint32_t> scopes;
+		// todo: windows and android
+
+#if HAVE_IFADDRS
+		struct ifaddrs *ifaddr = nullptr, *ifa = nullptr;
+		if (getifaddrs(&ifaddr) < 0) {
+		} else {
+			for (ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
+				if (!ifa->ifa_addr)
+					continue;
+				if (ifa->ifa_addr->sa_family != AF_INET6)
+					continue;
+
+				auto sa = *((struct sockaddr_in6 *)ifa->ifa_addr);
+			if (sa.sin6_scope_id)
+					scopes.push_back(sa.sin6_scope_id);
+
+				/*errorstream<<"in=" << ifa->ifa_name << " a="<<Address(*((struct sockaddr_in6*)ifa->ifa_addr)).serializeString()<<" ba=" << ifa->ifa_broadaddr <<" sc=" << sa.sin6_scope_id <<" fl=" <<  ifa->ifa_flags
+				//<< " bn=" << Address(*((struct sockaddr_in6*)ifa->ifa_broadaddr)).serializeString()
+				<<"\n"; */
+			}
+		}
+		freeifaddrs(ifaddr);
+#endif
+
+		if (scopes.empty())
+			scopes.push_back(0);
+
+		struct addrinfo hints
+		{
+		};
+
+		hints.ai_family = AF_INET6;
+		hints.ai_socktype = SOCK_DGRAM;
+		hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
+		struct addrinfo *result = nullptr;
+		if (!getaddrinfo("ff02::1", nullptr, &hints, &result)) {
+			for (auto info = result; info; info = info->ai_next) {
+				try {
+					sockaddr_in6 addr = *((struct sockaddr_in6 *)info->ai_addr);
+					addr.sin6_port = htons(adv_port);
+					UDPSocket socket_send(true);
+					int set_option_on = 1;
+					//setsockopt(socket_send.GetHandle(), SOL_SOCKET, IPV6_MULTICAST_HOPS,
+					//		(const char *)&set_option_on, sizeof(set_option_on));
+					auto use_scopes = scopes;
+					if (addr.sin6_scope_id) {
+						use_scopes.clear();
+						use_scopes.push_back(addr.sin6_scope_id);
+					}
+					for (auto &scope : use_scopes) {
+						addr.sin6_scope_id = scope;
+						socket_send.Send(Address(addr), str.c_str(), str.size());
+					}
+				} catch (const std::exception &e) {
+					verbosestream << "udp broadcast send over ipv6 fail " << e.what() << "\n";
+				}
+			}
+			freeaddrinfo(result);
+		}
+	}
+
+	try {
+		sockaddr_in addr = {};
+		addr.sin_family = AF_INET;
+		addr.sin_port = htons(adv_port);
+		UDPSocket socket_send(false);
+
+		// Full discloser, the use of inet_proto and the mreq stucture 
+		// were suggested by ai.
+
+		inet_pton(AF_INET, adv_multicast_addr, &(addr.sin_addr));
+
+		struct ip_mreq mreq;
+
+		mreq.imr_multiaddr.s_addr = inet_addr(adv_multicast_addr);
+		mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+		
+
+		setsockopt(socket_send.GetHandle(), IPPROTO_IP, IP_ADD_MEMBERSHIP,
+				(const char *)&mreq, sizeof(mreq));
```

Comment:
is should be moved into the socket class, if at all possible, you are breaking encapsulation here. 
Won't need to include all those OS specific headers once it's moved over.

Reply:
So, add code in the UDPSocket class constructer in socket.cpp to take another argument for the use of multicast, and add code to set up the socket for multicast, and then in the destructor, if we are using multicast then disconnect from that multicast group?

Reply:
I don't know what the best implementation in UDPSocket would be but what you suggested seems fine.

Reply:
It seems to me that it would be easiest to create a separate class that is a wrapper for the UDPSocket class that sets the correct socket options for multicast.